sw/device/mask_rom/boot.md

```diff
@@ -0,0 +1,110 @@
+# Pseudo-code for Secure Boot Process.
+
+This file should be read in conjunction with secure boot specification. Some
+references to that document are included.
+
+
+1. Power on (entirely in hardware) Open Q: Whether SW has to configure PMP
+  initial region.
+
+2. Execution Begins
+  - CRT Startup Code (ASM)
+    - Clean Slate (SRAM, Registers)
+    - Well-defined interrupt/exception handler
+        Open Q: Do we want to enable interrupts in mask rom?
+    - Jump into C code (`boot`):
+
+```c
+void boot(void) {
+  // - Clean all other hardware state (using DIF Reset functions)
+  clean_device_state();
+
+  // - Chip-specific startup functionality
+  //   - Clocks
+  //   - AST
+  //   - Flash
+  //   - Entropy
+  //   Open Q: Proprietary Software Strategy.
+  chip_specific_startup();
+
+  // - Read Boot Policy for ROM_EXT from Flash (2.b)
+  //   Open Q: How is this protected beyond being in flash?
+  boot_policy = read_boot_policy();
+
+  // - Determine which rom_ext slot is in use (2.b,2.c.i)
+  rom_ext_manifest = read_rom_ext_slot(boot_policy);
+
+  // Verify ROM_EXT Image Digest (2.c.ii)
+  if (!verify_rom_ext_hash(rom_ext_manifest)) {
+    // Boot Failure (check policy)
+  }
+
+  // Verify Signature (2.c.iii)
+  valid, signature = verify_rom_ext_signature(rom_ext_manifest);
```

Comment:
1. What selects the public key in `pub_key_selector(…)`? It must be the boot_info as we don't have a valid rom_ext yet?
2. One issue is there's things between this verify and `final_jump_to_rom_ext` that require rom_ext to be valid - certainly we've had this in relation to `peripheral_lockdown`, but also the root key derivation. How do we deal with that if we only validate at the time of the final jump?

Reply:
> 1. What selects the public key in `pub_key_selector(…)`? It must be the boot_info as we don't have a valid rom_ext yet?

We probably want enough flexibility to be able to manage multiple keys. The most common approach is to include a key id or public key in the manifest, so that the verifier can perform a lookup operation against the pub key table.  The verify operation is keyed by the mask_rom, so this is safe to do.

> 2. One issue is there's things between this verify and `final_jump_to_rom_ext` that require rom_ext to be valid - certainly we've had this in relation to `peripheral_lockdown`, but also the root key derivation. How do we deal with that if we only validate at the time of the final jump?

The preferred approach is to avoid having the mask rom derive any secrets. ROM should be limited to just locking the inputs to the key manager and other peripherals.  We are looking at ways in which the root_key derivation can be delayed even further in the boot flow when not needed for attestation purposes. This imposes additional isolation requirements in the key manager, which we are accounting for. 

Note that in this case root key is not a secret that is only visible to the boot_rom, so it is best to contain it within the scope of where it is consumed.